### ISSUE TITLE
A CLI-script does not produce the same results as the generated C# code.

### ISSUE THREAD

**Issue by CBrauer** [dotnet/machinelearning#5028]
Hey Guys,

I have found that the C# source code that is generated by the CLI script does not agree with the script output.

I have modified the solution that is generated by a cli-script to see it they agree.  The cli-script shows metric values that I cannot duplicate with the code that was generated by the script.

I believe that the MLModel.zip file is not for the optimal solution found by the script.
I have zipped everything into a file called “cli-script.zip” and I have put this Zip file on my GitHub account: https://github.com/CBrauer/Test-of-ML.NET.

The file “test.zip” contains the folder: “cli-script”.  This folder contains:
1.	Model2.  This is the modified project solution that was created by the cli-script.
2.	“run-cli-script-model2.bat”.  This runs script.
3.	screen.png.  This is a screen capture of the results of the script.
4.	verify.txt.  The output of the Model2 project.  This should give the same results as “screen.png”.
The script output does not produce the same results as the C# solution that was created by the script.

I, therefore, assume that “MLModel.zip” is not the correct model file.

Charles

**Comment by gvashishtha** [comment]
@justinormont I was able to reproduce this behavior. .7906 accuracy when training on Windows Subsystem for Linux, .7325 accuracy when running the generated code in Visual Studio. Is this behavior expected?

**Comment by justinormont** [comment]
Two issues: differing evaluation datasets, and leakage due to SMOTE.

The metrics printed in the CLI are evaluated on the validation dataset. The metrics calculated in the CodeGen'd code is evaluated on your test dataset.

Currently, the test dataset metrics are only reported from the CodeGen'd code, and not in the CLI.

Feature request -- The CLI and Model Builder should report test set metrics for the one final model when a test dataset is provided. /cc @JakeRadMSFT, @LittleLittleCloud

Your CLI command:
```bash
mlnet auto-train^
 --task binary-classification^
 --dataset "H:\HedgeTools\Datasets\rocket-train-classify-smote.csv"^
 --test-dataset "H:\HedgeTools\Datasets\rocket-test-classify.csv"^
 --label-column-name "Altitude"^
 --cache on^
 --has-header true^
 --name "Model2"^
 --max-exploration-time 28800
```

Since a validation dataset is not being provided, it is randomly split from your training dataset. 

Given the naming of your training dataset (`rocket-train-classify-smote.csv`), it looks like you're using SMOTE to upsample. You likely will do better on your final test metrics if you create a validation dataset. 

Currently you're leaking some information between rows as you're upsampling w/ SMOTE which creates near-duplicate rows of data. These rows are then split into a training and validation dataset. This causes some information about the validation dataset to be in the training dataset, which causes the validation metrics to be artificially high. More info -- https://en.wikipedia.org/wiki/Leakage_(machine_learning)#Training_example_leakage

Additional info about handling small datasets in ML․NET -- https://github.com/dotnet/machinelearning/issues/4028

Feature request -- Add a SMOTE transform to handle the process for the user. There's subtle things which can go wrong which our transform and its use in AutoML can avoid. /cc @harishsk

**Comment by CBrauer** [comment]
Thank you for your explanation.

**Comment by CBrauer** [comment]
Hi Justin,

 

First of all… shame on you. You work for Microsoft, and you should be running on Windows 10, like me.

In your previous email, you said SMOTE produces “data Leakage”. That is a new one on me.  I have never heard of that one before.  Besides, my training dataset was pre-processed, and one of the steps was to remove duplicate rows.  There are no duplicate rows in the training set.

 

The results I got was:

1.	Accuracy from script: 0.8933.  Accuracy from C# program: 0.8117.
2.	AUC from script: 0.9546.  Accuracy from C# program: 0.7190. from
3.	F! from script: 08937.  F1-score from C# program: 0.3894.

I do not use Accuracy as a reliable metric.  I’m more concerned about the AUC and F!-scores, and they are way off.

 

Charles

**Comment by CBrauer** [comment]
Hi Gopal,

 

First of all… shame on you. You work for Microsoft, and you should be running on Windows 10, like me.

 

In a previous email, Justin Ormont said SMOTE produces “data Leakage”. That is a new one on me.  I have never heard of that one before.  Besides, my training dataset was pre-processed, and one of the steps was to remove duplicate rows.  There are no duplicate rows in the training set.

 

I do not understand your accuracy numbers. The results I got was:

1.	Accuracy from script: 0.8933.  Accuracy from C# program: 0.8117.
2.	AUC from script: 0.9546.  Auc from C# program: 0.7190. from
3.	F!-score from script: 08937.  F1-score from C# program: 0.3894.

I do not use Accuracy as a reliable metric.  I’m more concerned about the AUC and F!-scores, and they are way off.

 

Charles
